3p/iframe-transport-client.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {tryParseJson} from '../src/json';
+import {dev, user} from '../src/log';
+import {IFRAME_TRANSPORT_EVENTS_TYPE} from '../src/iframe-transport-common';
+import {IframeMessagingClient} from './iframe-messaging-client';
+
+/** @private @const {string} */
+const TAG_ = 'iframe-transport-client';
+
+/**
+ * Receives event messages bound for this cross-domain iframe, from all
+ * creatives
+ */
+export class IframeTransportClient {
+
+  /** @param {!Window} win */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {?function(string,string)} */
+    this.listener_ = null;
+
+    /** @protected {!IframeMessagingClient} */
+    this.client_ = new IframeMessagingClient(win);
+    this.client_.setHostWindow(this.win_.parent);
+    this.client_.setSentinel(dev().assertString(
+        tryParseJson(this.win_.name)['sentinel'],
+        'Invalid/missing sentinel on iframe name attribute' + this.win_.name));
+    this.client_.makeRequest(
+        IFRAME_TRANSPORT_EVENTS_TYPE,
+        IFRAME_TRANSPORT_EVENTS_TYPE,
```

Comment:
I believe this makes sense. In the class at https://github.com/ampproject/amphtml/blob/master/src/iframe-helper.js#L412, subscribers send a message of a certain type. That is stored in this.clientWindows_
Later, when send() is called (line 446), it looks at this.clientWindows_ to see who wants messages of that type, and sends it to them.
I don't see how it could work with the message types being different. If you send a subscription request for message type X, how would it know to send you message type Y?